do/loadbalancers.go

```diff
@@ -179,6 +184,126 @@ func (l *loadbalancers) EnsureLoadBalancer(clusterName string, service *v1.Servi
 
 }
 
+func checkIfPortInRange(port int, fwPortRange string) bool {
+
+	fwPortList := strings.Split(fwPortRange, "-")
+	if fwPortList[0] == "all" {
+		return true
+	}
+	if len(fwPortList) == 1 {
+		fwPort, _ := strconv.Atoi(fwPortList[0])
+		if port == fwPort {
+			return true
+		}
+	} else {
+		fwLowerVal, _ := strconv.Atoi(fwPortList[0])
+		fwUpperVal, _ := strconv.Atoi(fwPortList[1])
+		if port >= fwLowerVal && port <= fwUpperVal {
+			return true
+		}
+	}
+
+	return false
+
+}
+
+func checkIfPortRequiresDeletion(port int, fwPortRange string) bool {
+
+	fwPortList := strings.Split(fwPortRange, "-")
+	if fwPortList[0] == "all" {
+		return false
+	}
+	if len(fwPortList) == 1 {
+		fwPort, _ := strconv.Atoi(fwPortList[0])
+		if port == fwPort {
+			return true
+		}
+	}
+
+	return false
+
+}
+
+func CreateFirewallsRuleRequest(sourcelb string, targetPort int) *godo.FirewallRulesRequest {
+	rr := &godo.FirewallRulesRequest{
+		InboundRules: []godo.InboundRule{
+			{
+				Protocol:  "tcp",
+				PortRange: strconv.Itoa(targetPort),
+				Sources: &godo.Sources{
+					LoadBalancerUIDs: []string{sourcelb},
+				},
+			},
+		},
+	}
+	return rr
+}
+
+type firewallRequest struct {
+	firewallId string
+	rules      *godo.FirewallRulesRequest
+}
+
+// EnsureFirewall ensures that the node firewalls have the correct rules for
+// the load balancer.
+//
+// EnsureFirewall will not modify service or nodes.
+func (l *loadbalancers) EnsureFirewall(lb *godo.LoadBalancer) error {
+	var ruleExists bool
+	var rulesToAdd []firewallRequest
+	var alreadyAdded bool
+	//var completedFirewalls []string
+
+	for _, dropletid := range lb.DropletIDs {
+		firewalls, _, err := l.client.Firewalls.ListByDroplet(context.TODO(), dropletid, nil)
+		if err != nil {
+			return fmt.Errorf("Error listing firewalls for droplet %d", dropletid)
+		}
+
+		if len(firewalls) > 0 {
+			for _, lbRule := range lb.ForwardingRules {
+				ruleExists = false
+				//check each firewall to see if rule already exists
+				for _, firewall := range firewalls {
+					for _, fwInboundRule := range firewall.InboundRules {
+						for _, fwSourcelbs := range fwInboundRule.Sources.LoadBalancerUIDs {
```

Comment:
This logic can be broken up into functions. Checking if the rule exists, deleting the rule, creating the rule.